### PR TITLE
proxmox*: add support for API tokens

### LINKED ISSUE #559: proxmox: allow to use a token for the API
##### SUMMARY
Instead of having to use a (security-sensitive) `api_password` parameter for the module, being able to use a token (cf https://pve.proxmox.com/wiki/User_Management#pveum_tokens) for the proxmox API would be nice. For now i resort to a dedicated cluster user with only `PVEVMAdmin` role..

the proxmoxer python module supports an API tokens since its 1.1.0 version

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
proxmox

### PR DESCRIPTION
**Pull request by Aversiste**
##### SUMMARY

Provides an alternative authentication method to the password.
The token validity is only tested during the first request after the instantiation of `ProxmoxAPI`, thus the presence of a call to `proxmox.version.get()` in `proxmox_template`.

Fixes #559, idea from @d4vydm.

Require testing for modules `proxmox` and `proxmox_template`.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

proxmox_kvm, proxmox_template, proxmox

##### ADDITIONAL INFORMATION

Example of use:

```
- name: Ensure test-vm is created
  proxmox_kvm:
    api_host: prx-test
    api_user: tleguern@pve
    api_token_id: api-pve-test
    api_token_secret: fad3db3d-b335-ee15-9fb0-ea1bf70128db
    node: prx-test-01
    name: test-vm
    ...
```

### CODE REVIEW

File: plugins/modules/cloud/misc/proxmox_kvm.py
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_kvm.py:53]:
```suggestion
    type: str
    version_added: 1.3.0
```
Same for the other.

I guess it's time for a doc fragment soon, and maybe some shared module_utils which provides the argspec for the doc fragment, and handles things like creating the `ProxmoxAPI` client? :)

Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_kvm.py:54]: Good idea, I will look into it.

File: plugins/modules/cloud/misc/proxmox_kvm.py
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_kvm.py:50]: This is one too much?

Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_kvm.py]: Fixed.

File: plugins/modules/cloud/misc/proxmox_kvm.py
Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_kvm.py:53]: Should I include the doc fragment in this PR or in another one ?

Comment (felixfontein) [reply plugins/modules/cloud/misc/proxmox_kvm.py:53]: Both is fine for me.

Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_kvm.py:53]: Ok, I will work on this in an other PR, probably next week.

**Comment by Aversiste** [comment]
Hi @landryb,

Can you check if this patch works for you ?

**Comment by landryb** [comment]
well that was my first experience with ansible 2.10 and collections, and oh boy it was .. convoluted. On an OpenBSD machine:
```
$doas pkg_add py3-virtualenv
$python3 -m virtualenv /tmp/ansible
$. /tmp/ansible/bin/activate
(ansible) $python3 -m pip install ansible
...
Successfully installed MarkupSafe-1.1.1 PyYAML-5.3.1 ansible-2.10.3 ansible-base-2.10.3 cffi-1.14.3 cryptography-3.2.1 jinja2-2.11.2 packaging-20.4 pycparser-2.20 pyparsing-2.4.7 six-1.15.0

(ansible) $python3 -m pip install proxmoxer requests
...
Successfully installed certifi-2020.11.8 chardet-3.0.4 idna-2.10 proxmoxer-1.1.1 requests-2.25.0 urllib3-1.26.2

(ansible) $ansible-galaxy collection install git+https://github.com/Aversiste/community.general,api-token
...
Installing 'community.general:1.3.0' to '/home/landry/.ansible/collections/ansible_collections/community/general'
(ansible) $export COLLECTIONS_PATHS=/home/landry/.ansible/collections
(ansible) $ansible-playbook -i hosts t.yml
```
but that fails with `"msg": "authorization on proxmox cluster failed with exception: __init__() got an unexpected keyword argument 'token_value'"` so my guess is that something is not using the right collection. strangely, with ansible-playbook -vvv it says
```
Using module file /home/landry/.ansible/collections/ansible_collections/community/general/plugins/modules/proxmox_kvm.py
```
so i'm puzzled as to why it doesnt work. support for api tokens was added in proxmoxer 1.1.1 and the remote production host is a proxmox 6.2 so supports api tokens too.

**Comment by landryb** [comment]
from looking a bit at the code, the exceptions is thrown by proxmoxer which apparently doesnt expect this arg, so the right codepath/collection file from this PR seems correctly used.

**Comment by landryb** [comment]
looking at proxmoxer/proxmoxer#20 it seems `api_token` was expected in the constructor, not `token_value` ? but this was renamed later on to expect `token_value` in proxmoxer/proxmoxer@0c4f3b0b, and that's what i have in the `backends/https.py` file for the proxmoxer 1.1.1 installed by pip. With what version of proxmoxer did you test this PR ?

**Comment by Aversiste** [comment]
Due to a specific network separation at work I am not able to access proxmox servers from my OpenBSD laptop but from Ubuntu I was able to test this PR with proxmoxer versions 1.0.3, 1.0.4, 1.1.0, 1.1.1 and 1.2.0b1: the last three work as intended.
The error message `__init__() got an unexpected keyword argument 'token_value'` only appears with version priors to 1.1.0.

**Comment by landryb** [comment]
well that's puzzling because in the same venv, locally testing proxmoxer with the token works fine:
```
(ansible) [18:22] adventurer:/tmp/ansible/ $python                                                                                                 
Python 3.8.6 (default, Oct 30 2020, 16:02:47) 
>>> from proxmoxer import ProxmoxAPI
>>> t=ProxmoxAPI('umbrail', verify_ssl=False, user='user@pam', token_name='ansible', token_value='xxxx')
>>> t.version.get()
{'repoid': '48bd51b6', 'version': '6.2-15', 'release': '6.2'}
```

wierd... but maybe that's because i have a frankeinstein install now in that venv.. but still, i only have one proxmoxer version installed on that system.

**Comment by Aversiste** [comment]
Then I guess this is not the `proxmoxer` ansible is loading. One way to be sure is to try out this patch:

```
diff --git a/plugins/modules/cloud/misc/proxmox_kvm.py b/plugins/modules/cloud/misc/proxmox_kvm.py
index 9d9d1518..b5eb7f7e 100644
--- a/plugins/modules/cloud/misc/proxmox_kvm.py
+++ b/plugins/modules/cloud/misc/proxmox_kvm.py
@@ -718,10 +718,13 @@ from distutils.version import LooseVersion
 
 try:
     from proxmoxer import ProxmoxAPI
+    from proxmoxer import __version__ as ProxmoxVersion
     HAS_PROXMOXER = True
 except ImportError:
     HAS_PROXMOXER = False
 
+raise Exception(ProxmoxVersion)
+
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils._text import to_native
```

**Comment by landryb** [comment]
Bah. figured out my failure, with my test playbook the proxmoxer used is the one on the remote node (which is the debian stable python-proxmoxer 1.0.3 package) and not the one i installed locally via pip.

once i use `api_host: remote_host` instead of `api_host: localhost` and run the playbook on the localhost (ie the openbsd laptop with proxmoxer 1.1.0 installed via pip) i can successfully communicate with the proxmox on `remote_host`, using the API token. For some reason it cant find the VM im trying to get the state of, but i see correctly authentified requests in `/var/log/pveproxy/access.log`.

so the PR works for me - thanks !

**Comment by Aversiste** [comment]
Great :)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@Aversiste do you want to do more changes / wait for more reviews/testing, or should I merge?

**Comment by Aversiste** [comment]
@felixfontein I believe it can be merged as is :)

**Comment by felixfontein** [comment]
@Averiste thanks for implementing this!
@landryb thanks for testing!

**Comment by landryb** [comment]
> @Averiste thanks for implementing this!
> @landryb thanks for testing!

that's the true power of open source after all ;) thanks for merging !

**Comment by d4vydm** [comment]
Awesome work!

**Pull request merged**
